kernel/src/capability_ptr.rs

```diff
@@ -0,0 +1,135 @@
+// Licensed under the Apache License, Version 2.0 or the MIT License.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// Copyright Google LLC 2024.
+
+//! Defines the CapabilityPtr type
+
+use core::fmt::{Formatter, LowerHex, UpperHex};
+use core::ops::AddAssign;
+
+/// A pointer to userspace memory.
+///
+/// A [CapabilityPtr] points to memory a userspace processes may be
+/// permitted to read, write, or execute. It is sized exacly to a
+/// register that can pass values between userspace and kernel and at
+/// least the size of a word ([usize]) [^note1]. Operations on the
+/// pointer may affect permissions, e.g. offsetting the pointer beyond
+/// the bounds of the memory object invalidates it. Like a `*const
+/// ()`, a [CapabilityPtr] may also "hide" information by storing a
+/// word of data with no memory access permissions.
+///
+/// [CapabilityPtr] should be used to store or pass between kernel and
+/// userspace a value that may represent a valid userspace reference.
+///
+/// [^note1]: Depending on the architecutre, this the size of a
+/// [CapabilityPtr] may be a word size or larger, e.g., if registers
+/// can store metadata such as access permissions.
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
+#[repr(transparent)]
+pub struct CapabilityPtr {
+    ptr: *const (),
+}
+
+impl Default for CapabilityPtr {
+    fn default() -> Self {
+        Self {
+            ptr: core::ptr::null(),
+        }
+    }
+}
+
+#[derive(Copy, Clone, PartialEq)]
+pub enum MetaPermissions {
+    Any,
+    Read,
+    Write,
+    ReadWrite,
+    Execute,
+}
+
+impl From<CapabilityPtr> for usize {
+    #[inline]
+    fn from(from: CapabilityPtr) -> Self {
+        from.ptr as usize
+    }
+}
+
+impl From<usize> for CapabilityPtr {
+    #[inline]
+    fn from(from: usize) -> Self {
+        Self {
+            ptr: from as *const (),
+        }
+    }
+}
+
+impl UpperHex for CapabilityPtr {
+    #[inline]
+    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
+        UpperHex::fmt(&(self.ptr as usize), f)
+    }
+}
+
+impl LowerHex for CapabilityPtr {
+    #[inline]
+    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
+        LowerHex::fmt(&(self.ptr as usize), f)
+    }
+}
+
+impl AddAssign<usize> for CapabilityPtr {
+    #[inline]
+    fn add_assign(&mut self, rhs: usize) {
+        self.ptr = (self.ptr as *const u8).wrapping_add(rhs) as *const ();
+    }
+}
+
+impl CapabilityPtr {
+    pub fn as_ptr<T>(&self) -> *const T {
+        self.ptr as *const T
+    }
+
+    /// Convert to a raw pointer, checking that metadata allows a particular set of permissions over
+    /// a given number of bytes.
+    /// If the metadata does not allow for this, returns null.
+    /// If no such metadata exists, this succeeds.
+    #[inline]
+    pub fn as_ptr_checked<T>(&self, _length: usize, _perms: MetaPermissions) -> *const T {
```

Comment:
I feel somewhat strongly that this should return an `Option<*const T>` in favor of the implicit coercion to null. This is a difference in semantics that we want to be explicit about, for instance in the cases where it's used in `syscall.rs` (as pointed out by @jrvanwhy): https://github.com/tock/tock/pull/4174/files#r1828604067

Reply:
I think that plan works, with special consideration about allowing NULL to be passed. Removing this function from this PR entirely, however, as discussed.